libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    public class ShowTypingMiddleware : IMiddleware
+    {
+        /// <summary>
+        /// (Optional) initial delay before sending first typing indicator. Defaults to 500ms.
+        /// </summary>
+        private readonly int _delay;
+
+        /// <summary>
+        /// (Optional) rate at which additional typing indicators will be sent. Defaults to every 2000ms.
+        /// </summary>
+        private readonly int _freqency;
+
+        public ShowTypingMiddleware(int delay = 500, int frequency = 2000)
+        {
+            if(delay < 0)
+                throw new ArgumentOutOfRangeException(nameof(delay), "Delay must be greater than or equal to zero");
+
+            if (frequency <= 0)
+                throw new ArgumentOutOfRangeException(nameof(frequency), "Frequency must be greater than zero");
+
+            _delay = delay;
+            _freqency = frequency;
+        }
+
+        public async Task OnProcessRequest(IBotContext context, MiddlewareSet.NextDelegate next)
+        {
+            Timer typingActivityTimer = null;
+
+            // If the incoming activity is a MessageActivity, start a timer to periodically send the typing activity
+            if (context.Request.Type == ActivityTypes.Message)
+            {
+                typingActivityTimer = new Timer(SendTypingTimerCallback, context, _delay, _freqency);
+            }
+
+            await next().ConfigureAwait(false);
```

Comment:
So if an exception bubble up here you will "leak" the timer. Suggest adding a `try/finally` around this to ensure you always dispose of the `typingActivityTimer`.